unit_tests/sources/declarative/test_state_delegating_stream.py

```diff
@@ -0,0 +1,248 @@
+        stream=stream,
+        sync_mode=sync_mode,
+        destination_sync_mode=destination_sync_mode,
+        cursor_field=cursor_field,
+        primary_key=primary_key,
+    )
+
+
+def to_configured_catalog(
+    configured_streams,
+) -> ConfiguredAirbyteCatalog:
+    return ConfiguredAirbyteCatalog(streams=configured_streams)
+
+
+def create_configured_catalog(
+    source: ConcurrentDeclarativeSource, config: dict
+) -> ConfiguredAirbyteCatalog:
+    """
+    Discovers streams from the source and converts them into a configured catalog.
+    """
+    actual_catalog = source.discover(logger=source.logger, config=config)
+    configured_streams = [
+        to_configured_stream(stream, primary_key=stream.source_defined_primary_key)
+        for stream in actual_catalog.streams
+    ]
+    return to_configured_catalog(configured_streams)
+
+
+def get_records(
+    source: ConcurrentDeclarativeSource,
+    config: dict,
+    catalog: ConfiguredAirbyteCatalog,
+    state: list = None,
+) -> list:
+    """
+    Reads records from the source given a configuration, catalog, and optional state.
+    Returns a list of record data dictionaries.
+    """
+    return [
+        message.record.data
+        for message in source.read(logger=MagicMock(), config=config, catalog=catalog, state=state)
+        if message.type == Type.RECORD
+    ]
+
+
+@freezegun.freeze_time("2024-07-15")
+def test_state_retriever():
+    with HttpMocker() as http_mocker:
+        http_mocker.get(
+            HttpRequest(url="https://api.test.com/items"),
+            HttpResponse(
+                body=json.dumps(
+                    [
+                        {"id": 1, "name": "item_1", "updated_at": "2024-07-13"},
+                        {"id": 2, "name": "item_2", "updated_at": "2024-07-13"},
+                    ]
+                )
+            ),
+        )
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.test.com/items_with_filtration?start=2024-07-13&end=2024-07-15"
+            ),
+            HttpResponse(
+                body=json.dumps(
+                    [
+                        {"id": 3, "name": "item_3", "updated_at": "2024-02-01"},
+                        {"id": 4, "name": "item_4", "updated_at": "2024-02-01"},
+                    ]
+                )
+            ),
+        )
+
+        source = ConcurrentDeclarativeSource(
+            source_config=_MANIFEST, config=_CONFIG, catalog=None, state=None
+        )
+        configured_catalog = create_configured_catalog(source, _CONFIG)
+
+        # Test full data retrieval (without state)
+        full_records = get_records(source, _CONFIG, configured_catalog)
+        expected_full = [
+            {"id": 1, "name": "item_1", "updated_at": "2024-07-13"},
+            {"id": 2, "name": "item_2", "updated_at": "2024-07-13"},
+        ]
+        assert expected_full == full_records
+
+        # Test incremental data retrieval (with state)
```

Comment:
nit: can we split this in two tests? It is interesting feedback for the users to see that full refresh is failing but not incremental. With the current test setup, it is not possible to know

Reply:
Done